File: mcp-samples/02-sample-code-gen/code-gen-source-samples/javascript/controllers/userController.js
/**
 * User controller module
 *
 * Defines Express routes and handlers for user operations
 */

const express = require('express');

/**
 * Creates an Express router with user endpoints
 * @param {UserService} userService - User service instance
 * @returns {express.Router} Configured Express router
 */
function createUserController(userService) {
  const router = express.Router();

  /**
   * GET /api/users
   * Get all users
   */
  router.get('/', async (req, res) => {
    try {
      const users = await userService.getAll();
      res.status(200).json(users);
    } catch (error) {
      console.error('Error getting all users:', error);
      res.status(500).json({
        message: 'Error retrieving users',
        error: error.message
      });
    }
  });

  /**
   * GET /api/users/:id
   * Get user by ID
   */
  router.get('/:id', async (req, res) => {
    try {
      const id = parseInt(req.params.id, 10);
      if (isNaN(id)) {
        return res.status(400).json({
          message: 'Invalid user ID'
        });
      }

      const user = await userService.getById(id);
      res.status(200).json(user);
    } catch (error) {
      console.error(`Error getting user ${req.params.id}:`, error);

      if (error.message.includes('not found')) {
        return res.status(404).json({
          message: error.message
        });
      }

      res.status(500).json({
        message: 'Error retrieving user',
        error: error.message
      });
    }
  });

  /**
   * POST /api/users
   * Create a new user
   */
  router.post('/', async (req, res) => {
    try {
      const userData = req.body;

      // Basic validation
      if (!userData || typeof userData !== 'object') {
        return res.status(400).json({
          message: 'Invalid request body'
        });
      }

      const user = await userService.create(userData);
      res.status(201).json(user);
    } catch (error) {
      console.error('Error creating user:', error);

      if (error.message.includes('Validation failed') ||
          error.message.includes('already exists')) {
        return res.status(400).json({
          message: error.message
        });
      }

      res.status(500).json({
        message: 'Error creating user',
        error: error.message
      });
    }
  });

  /**
   * PUT /api/users/:id
   * Update an existing user
   */
  router.put('/:id', async (req, res) => {
    try {
      const id = parseInt(req.params.id, 10);
      if (isNaN(id)) {
        return res.status(400).json({
          message: 'Invalid user ID'
        });
      }

      const userData = req.body;
      if (!userData || typeof userData !== 'object') {
        return res.status(400).json({
          message: 'Invalid request body'
        });
      }

      const user = await userService.update(id, userData);
      res.status(200).json(user);
    } catch (error) {
      console.error(`Error updating user ${req.params.id}:`, error);

      if (error.message.includes('not found')) {
        return res.status(404).json({
          message: error.message
        });
      }

      if (error.message.includes('Validation failed') ||
          error.message.includes('already in use')) {
        return res.status(400).json({
          message: error.message
        });
      }

      res.status(500).json({
        message: 'Error updating user',
        error: error.message
      });
    }
  });

  /**
   * DELETE /api/users/:id
   * Delete a user
   */
  router.delete('/:id', async (req, res) => {
    try {
      const id = parseInt(req.params.id, 10);
      if (isNaN(id)) {
        return res.status(400).json({
          message: 'Invalid user ID'
        });
      }

      await userService.delete(id);
      res.status(204).send();
    } catch (error) {
      console.error(`Error deleting user ${req.params.id}:`, error);

      if (error.message.includes('not found')) {
        return res.status(404).json({
          message: error.message
        });
      }

      res.status(500).json({
        message: 'Error deleting user',
        error: error.message
      });
    }
  });

  /**
   * GET /api/users/email/:email
   * Find user by email
   */
  router.get('/email/:email', async (req, res) => {
    try {
      const email = req.params.email;
      if (!email) {
        return res.status(400).json({
          message: 'Email parameter is required'
        });
      }

      const user = await userService.findByEmail(email);
      if (!user) {
        return res.status(404).json({
          message: `User with email ${email} not found`
        });
      }

      res.status(200).json(user);
    } catch (error) {
      console.error(`Error finding user by email ${req.params.email}:`, error);
      res.status(500).json({
        message: 'Error searching for user',
        error: error.message
      });
    }
  });

  return router;
}

module.exports = { createUserController };

File: mcp-samples/02-sample-code-gen/code-gen-source-samples/dotnet/controllers/UserController.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Mvc;
using YourProject.Models;
using YourProject.Services;

namespace YourProject.Controllers
{
    /// <summary>
    /// REST API controller for User operations
    /// </summary>
    [ApiController]
    [Route("api/[controller]")]
    public class UserController : ControllerBase
    {
        private readonly IUserService _userService;

        public UserController(IUserService userService)
        {
            _userService = userService;
        }

        /// <summary>
        /// Gets all users
        /// </summary>
        /// <returns>List of users</returns>
        [HttpGet]
        public async Task<ActionResult<IEnumerable<User>>> GetAll()
        {
            try
            {
                var users = await _userService.GetAllAsync();
                return Ok(users);
            }
            catch (Exception ex)
            {
                return StatusCode(500, new { message = "Error retrieving users", error = ex.Message });
            }
        }

        /// <summary>
        /// Gets a user by ID
        /// </summary>
        /// <param name="id">User ID</param>
        /// <returns>User details</returns>
        [HttpGet("{id}")]
        public async Task<ActionResult<User>> GetById(int id)
        {
            try
            {
                var user = await _userService.GetByIdAsync(id);
                return Ok(user);
            }
            catch (KeyNotFoundException ex)
            {
                return NotFound(new { message = ex.Message });
            }
            catch (Exception ex)
            {
                return StatusCode(500, new { message = "Error retrieving user", error = ex.Message });
            }
        }

        /// <summary>
        /// Creates a new user
        /// </summary>
        /// <param name="user">User details</param>
        /// <returns>Created user</returns>
        [HttpPost]
        public async Task<ActionResult<User>> Create([FromBody] User user)
        {
            try
            {
                if (!ModelState.IsValid)
                {
                    return BadRequest(ModelState);
                }

                var createdUser = await _userService.CreateAsync(user);
                return CreatedAtAction(nameof(GetById), new { id = createdUser.Id }, createdUser);
            }
            catch (ArgumentException ex)
            {
                return BadRequest(new { message = ex.Message });
            }
            catch (Exception ex)
            {
                return StatusCode(500, new { message = "Error creating user", error = ex.Message });
            }
        }

        /// <summary>
        /// Updates an existing user
        /// </summary>
        /// <param name="id">User ID</param>
        /// <param name="user">Updated user details</param>
        /// <returns>Updated user</returns>
        [HttpPut("{id}")]
        public async Task<ActionResult<User>> Update(int id, [FromBody] User user)
        {
            try
            {
                if (!ModelState.IsValid)
                {
                    return BadRequest(ModelState);
                }

                var updatedUser = await _userService.UpdateAsync(id, user);
                return Ok(updatedUser);
            }
            catch (KeyNotFoundException ex)
            {
                return NotFound(new { message = ex.Message });
            }
            catch (Exception ex)
            {
                return StatusCode(500, new { message = "Error updating user", error = ex.Message });
            }
        }

        /// <summary>
        /// Deletes a user
        /// </summary>
        /// <param name="id">User ID</param>
        /// <returns>Success status</returns>
        [HttpDelete("{id}")]
        public async Task<ActionResult> Delete(int id)
        {
            try
            {
                await _userService.DeleteAsync(id);
                return NoContent();
            }
            catch (KeyNotFoundException ex)
            {
                return NotFound(new { message = ex.Message });
            }
            catch (Exception ex)
            {
                return StatusCode(500, new { message = "Error deleting user", error = ex.Message });
            }
        }
    }
}
